File: jsrc/kernelprolog/fluents/SxxMachine/StringSink.java
package SxxMachine;

/**
  Builds  Fluents from Java
  Streams
*/
public class StringSink extends Sink {
	protected StringBuffer buffer;

	public StringSink(Prog p) {
		super(p);
		this.buffer = new StringBuffer();
	}

	public int putElement(Term t) {
		buffer.append(t.toUnquoted());
		return 1;
	}

	public void stop() {
		buffer = null;
	}

	public Term collect() {
		return TermData.SYM(buffer.toString());
	}
}

File: src/kernelprolog/fluents/SxxMachine/StringSink.cpp
using namespace std;

#include "StringSink.h"
#include "../../terms/SxxMachine/Prog.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../main/SxxMachine/Builtins.h"
#include "../../../machine/SxxMachine/TermData.h"
#include "StringBuilder.h"

namespace SxxMachine
{

	StringSink::StringSink(Prog* p) : Sink(p)
	{
		this->buffer = new StringBuilder();
	}

	int StringSink::putElement(Term* t)
	{
		buffer->append(t->toUnquoted());
		return 1;
	}

	void StringSink::stop()
	{
		buffer = nullptr;
	}

	Term* StringSink::collect()
	{
		return TermData::SYM(buffer->toString());
	}
}
